internal/xds/resolver/xds_http_filters_test.go

```diff
@@ -0,0 +1,678 @@
+type testFilterCfg struct {
+	httpfilter.FilterConfig
+	path         string
+	newStreamErr error
+}
+
+// filterConfigFromProto parses filter config specified as a v3.TypedStruct into
+// a testFilterCfg.
+func filterConfigFromProto(cfg proto.Message) (httpfilter.FilterConfig, error) {
+	ts, ok := cfg.(*v3xdsxdstypepb.TypedStruct)
+	if !ok {
+		return nil, fmt.Errorf("unsupported filter config type: %T, want %T", cfg, &v3xdsxdstypepb.TypedStruct{})
+	}
+
+	if ts.GetValue() == nil {
+		return testFilterCfg{}, nil
+	}
+	ret := testFilterCfg{}
+	if v := ts.GetValue().GetFields()[filterCfgPathFieldName]; v != nil {
+		ret.path = v.GetStringValue()
+	}
+	if v := ts.GetValue().GetFields()[filterCfgErrorFieldName]; v != nil {
+		if v.GetStringValue() == "" {
+			ret.newStreamErr = nil
+		} else {
+			ret.newStreamErr = fmt.Errorf("%s", v.GetStringValue())
+		}
+	}
+	return ret, nil
+}
+
+type logger interface {
+	Logf(format string, args ...any)
+}
+
+// testHTTPFilterWithRPCMetadata is a HTTP filter used for testing purposes.
+//
+// This filter is used to verify that the xDS resolver and filter stack
+// correctly propagate filter configuration (both base and override) to RPCs. It
+// does this by injecting the config paths from its base and override configs as
+// JSON-encoded metadata into outgoing RPCs.  The metadata can then be observed
+// by the backend, allowing tests to assert that the correct filter
+// configuration was applied for each RPC.
+type testHTTPFilterWithRPCMetadata struct {
+	logger        logger
+	typeURL       string
+	newStreamChan *testutils.Channel // If set, filter config is written to this field from NewStream()
+}
+
+func (fb *testHTTPFilterWithRPCMetadata) TypeURLs() []string { return []string{fb.typeURL} }
+
+func (*testHTTPFilterWithRPCMetadata) ParseFilterConfig(cfg proto.Message) (httpfilter.FilterConfig, error) {
+	return filterConfigFromProto(cfg)
+}
+
+func (*testHTTPFilterWithRPCMetadata) ParseFilterConfigOverride(override proto.Message) (httpfilter.FilterConfig, error) {
+	return filterConfigFromProto(override)
+}
+
+func (*testHTTPFilterWithRPCMetadata) IsTerminal() bool { return false }
+
+var _ httpfilter.ClientInterceptorBuilder = &testHTTPFilterWithRPCMetadata{}
+
+func (fb *testHTTPFilterWithRPCMetadata) BuildClientInterceptor(config, override httpfilter.FilterConfig) (iresolver.ClientInterceptor, error) {
+	fb.logger.Logf("BuildClientInterceptor called with config: %+v, override: %+v", config, override)
+
+	if config == nil {
+		panic("unexpected missing config")
+	}
+
+	baseCfg := config.(testFilterCfg)
+	basePath := baseCfg.path
+	newStreamErr := baseCfg.newStreamErr
+
+	var overridePath string
+	if override != nil {
+		overrideCfg := override.(testFilterCfg)
+		overridePath = overrideCfg.path
+		if overrideCfg.newStreamErr != nil {
+			newStreamErr = overrideCfg.newStreamErr
+		}
+	}
+	var errStr string
+	if newStreamErr != nil {
+		errStr = newStreamErr.Error()
```

Comment:
I see that we're converting a string to an error, assigning it to `testFilterCfg.newStreamErr` and converting it back to a string here. Can the error string be stored in `testFilterCfg.newStreamErr` to avoid these conversions?

Reply:
Done.